### ISSUE TITLE
False positive for `into_iter` rule

### ISSUE THREAD

**Issue by flying-sheep** [Manishearth/rust-clippy#1398]
Calling [`Itertools::chunks`](https://docs.rs/itertools/*/itertools/trait.Itertools.html#method.chunks) returns a [`IntoChunks`](https://docs.rs/itertools/*/itertools/structs/struct.IntoChunks.html) with `impl<'a> IntoIterator for &'a IntoChunks`

```rust
for records_iter in some_iterator.chunks().into_iter() { ... }
```

works while

```rust
for records_iter in some_iterator.chunks() { ... }
```

fails with

```rust
error[E0277]: the trait bound `itertools::IntoChunks<T>: std::iter::Iterator` is
 not satisfied
[…]
    = note: `itertools::IntoChunks<T>` is not an iterator; maybe try calling `.i
ter()` or a similar method
    = note: required by `std::iter::IntoIterator::into_iter`
```

**Comment by oli-obk** [comment]
That makes no sense in my book... `Vec<T>` doesn't implement `Iterator` and it works just fine through the `IntoIterator` impl. Have you tried reporting this to `itertools`?

**Comment by mcarton** [comment]
Actually it should be `for records_iter in &some_iterator.chunks(…)` (see [`IntoChunks`](https://docs.rs/itertools/*/itertools/structs/struct.IntoChunks.html)). Don't know why, it seems like some inconvenience I would report to `itertools` too.
(an explicit call to `into_iter` works without the `&` because a method call can add one `&` if it needs to)

The lint should however not trigger anyway.

**Comment by flying-sheep** [comment]
thanks, makes sense!

**Comment by bluss** [comment]
In fact @mcarton linked to docs for IntoChunks that explain why. Maybe arrays have the same false positive? `let array = [1, 2, 3]; for elt in array.into_iter()` is the same kind of thing.

**Comment by bluss** [comment]
Great.

Come to think of it, maybe clippy should lint against the `[0, 1, 2].into_iter()` case since it is a very tricky area — a by value iterator for arrays would like to change meaning for that expression. Of course I'm by no means sure that will happen or if it is feasible to do such a breaking change.
